packages/shared/src/utils/transactions.ts

```diff
@@ -183,9 +216,73 @@ export const getGasRefillChainedTxParams = (
   }
 ]
 
+export const getMissingBalancesChainedTxParams = (
+  signerMissingBalances: Map<string, bigint>,
+  addressWithEnoughBalances: AddressWithGroup,
+  signerAddress: AddressHash,
+  txParams: TransactionParams
+): Array<SignChainedTxParams> => {
+  const tokens = Array.from(signerMissingBalances.entries())
+    .filter(([tokenId]) => tokenId !== ALPH.id)
+    .map(([tokenId, amount]) => ({ id: tokenId, amount: amount.toString() }))
+  const attoAlphAmount = signerMissingBalances.get(ALPH.id)?.toString() || DUST_AMOUNT * BigInt(tokens.length)
+
+  return [
+    {
+      type: 'Transfer',
+      signerAddress: addressWithEnoughBalances.hash,
+      signerKeyType: addressWithEnoughBalances.keyType,
+      destinations: [{ address: signerAddress, attoAlphAmount, tokens }]
+    },
+    transactionParamsToSignChainedTxParams(txParams)
+  ]
+}
+
 export const getSweepTxParams = (data: SendFlowData): SweepTxParams => ({
   signerAddress: data.fromAddress.hash,
   signerKeyType: data.fromAddress.keyType,
   toAddress: data.toAddress,
   lockTime: data.lockTime?.getTime()
 })
+
+export const getTransactionExpectedBalances = ({ type, params }: TransactionParams) => {
+  const expectedBalances: Map<string, bigint> = new Map()
+
+  const addTokenToExpectedBalances = (tokenId: string, amount: bigint) => {
+    const prevValue = expectedBalances.get(tokenId) ?? BigInt(0)
+    expectedBalances.set(tokenId, prevValue + amount)
```

Comment:
Interesting way of doing this. Reminds me of a few years ago, when I was doing more functional programming. I like it!